docs/src/architecture/notes.md

```diff
@@ -1,73 +1,194 @@
+          # => [ASSET, ptr, end_ptr, ...]
+  
+          # increment the pointer and compare it to the end_ptr
+          movup.4 add.1 dup dup.6 neq
+          # => [latch, ptr+1, ASSET, end_ptr, ...]
+      end
+  
+      # clear the stack
+      drop dropw drop
+  end
+  ```
+
+  The procedure starts by calling `exec.note::get_assets` and putting the note's number of assets and the memory pointer of the first asset on top of the stack. Assets are stored in consecutive memory slots, so `dup.1 add` provides the last memory slot. Because [assets](assets.md) are represented by `Words` in Miden Assembly, the procedure pads the stack with four `0`s. Now, if there is at least one asset (checked by `dup dup.6 neq`), the loop starts. It first saves the pointer for later use (`dup movdn.5`), then loads the first asset `mem_loadw` on top of the stack. Now, the procdure calls the a function of the account interface `call.wallet::receive_asset` to put the asset into the account's vault. The note script cannot directly call an account function to add the asset. The account must expose this function in its interface. Lastly, the pointer gets incremented, and if there is a second asset, the loop continues (`movup.4 add.1 dup dup.6 neq`). Finally, when all assets were put into the account's vault, the stack is cleared (`drop dropw drop`).
+
+</details>
+
+## Note storage mode
+Similar to accounts, there are two storage modes for notes in Miden. Notes can be stored publicly in the [Notes DB](https://0xpolygonmiden.github.io/miden-base/architecture/state.html#notes-database) with all data visible for everyone. Alternatively, notes can be stored privately committing only the note hash to the Notes DB. 
+
+Every note has a unique single note hash. It is defined as `hash(hash(hash(hash(serial_num, [0; 4]), script_hash), input_hash), vault_hash)`. It is easy to compute the note hash if all the note data is known. To compute a note's hash, we do not need to know the note's `serial_num`. Knowing the hash of the `serial_num` (as well as `script_hash`, `input_hash` and `note_vault`) is also sufficient. We compute the hash of `serial_num` as `hash(serial_num, [0; 4])` to simplify processing within the VM.
+
+Privately stored notes can only be consumed if the note data is known to the consumer. The note data must be provided to the [transaction kernel](../transactions/transaction-kernel.md). That means, there must be some offchain communication to transmit the note's data from the sender to the recipient.
+
+## Note discovery
+Note discovery describes the process of Miden Clients finding notes they want to consume. It will always be possible to send note data directly to the target account, and if the note was previously recored on-chain, it can be consumed. However, it is also possible to querry the Miden Operator and request newly recorded relevant notes. This is done via Note Tags. Tags are part of the Note's metadata and are represented by a `Felt`. 
```

Comment:
Is it only Miden Clients or more broadly the process of finding notes that users want to consume.

Unclear: It will always be possible to send note data directly to the target account, and if the note was previously recored on-chain, it can be consumed.

What is the Miden Operator? I would change this to "rollup", "network".

I would change relevant to matching certain set requirements which is done via Note Tags.

At the end I would add information on "why tags",  tags are useful for note discovery enabling easy collection of all notes matching a certain tag. Or something like this.

Reply:
> Is it only Miden Clients or more broadly the process of finding notes that users want to consume.

Let's start with Miden clients. 

I changed the text